Src/java/cql-to-elm/src/main/java/org/cqframework/cql/cql2elm/Cql2ElmVisitor.java

```diff
@@ -242,12 +301,142 @@ private void popChunk(@NotNull ParseTree tree, Object o) {
         }
     }
 
+    private void processTags(@NotNull ParseTree tree, Object o) {
+        if (o instanceof Element) {
+            Element element = (Element)o;
+            if (!(tree instanceof cqlParser.LibraryContext)) {
+                if (element instanceof UsingDef || element instanceof IncludeDef || element instanceof CodeSystemDef
+                        || element instanceof ValueSetDef || element instanceof CodeDef || element instanceof ConceptDef
+                        || element instanceof ParameterDef || element instanceof ContextDef || element instanceof ExpressionDef) {
+                    List<Tag> tags = getTags(tree);
+                    if (tags != null && tags.size() > 0) {
+                        Annotation a = getAnnotation(element);
+                        if (a == null) {
+                            a = buildAnnotation();
+                            element.getAnnotation().add(a);
+                        }
+                        a.getT().addAll(tags);
+                    }
+                }
+            }
+            else {
+                if (libraryInfo.getDefinition() != null && libraryInfo.getHeaderInterval() != null) {
+                    List<Tag> tags = getTags(libraryInfo.getHeader());
+                    if (tags != null && tags.size() > 0) {
+                        Annotation a = getAnnotation(libraryBuilder.getLibrary());
+                        if (a == null) {
+                            a = buildAnnotation();
+                            libraryBuilder.getLibrary().getAnnotation().add(a);
+                        }
+                        a.getT().addAll(tags);
+                    }
+                }
+            }
+        }
+    }
+
+    private List<Tag> getTags(ParseTree tree) {
+        BaseInfo bi = libraryInfo.resolveDefinition(tree);
+        if (bi != null) {
+            return getTags(bi.getHeader());
+        }
+
+        return null;
+    }
+
+    private List<Tag> parseTags(String header) {
```

Comment:
I understand that the Tag grammar was introduced inside of the Comment grammar to facilitate backwards compatibility with 1.4. I think it'd be worthwhile to consider an alternative approach where there's a well-defined API for add processors to the Translator to allow injection of arbitrary pre or comment processing. That way the specification for tags could evolve separately from CQL itself, and defines an extension point for the CQL language in the event someone needs implementation specific functionality (for example, a comment grammar that allows binding a CQL definition to a precomputed dataset with a syntax that may not align well with Tags).

This approach is not unheard of with other languages such as Java. The javadoc processor is actually external to the java compiler and comments are stripped from bytecode.

I see that tags are trial-use for 1.5 so I may make a comment to that effect on spec itself.